Docs: Rewriting Extensions key concepts (#2030)
grafana/plugin-tools: docusaurus/docs/key-concepts/ui-extensions.md

---
id: ui-extensions

> title: Extensions

Reviewer: I personally don't mind this change, however if I remember correctly @sympatheticmoose and @tolzhabayev were against calling it only "Extensions", due to it being more ambiguous. What is our stance on it?

Author: We could rename it as "Functionality extensions" or similar. WDYT?

Reviewer 2: https://docs.google.com/document/d/1VYI4Do8663Q4VqhomqQksp2Y-HnznKK4nY122xvKZZ0/edit?disco=AAABoi4FwBc

I vote to keep the naming as is, not because it is ideal but because I don't think the other naming proposals we have so far are better.

Reviewer 3: Agree with Timur. If we want to rename, we can tackle that separately to this

  - links
  - extensions
  - app plugins
  - extension points
  - components
  - content
sidebar_position: 60
---

Use extensions to contribute new actions and functionality to the core Grafana UI and other app plugins. 

## Understand the extensions ecosystem

The UI extensions framework is built around these concepts:

> - **Extension point**: A place in Grafana Core or in a plugin where content can be rendered via hooks.

Reviewer: Would we like to mention hooks (React hooks) here (which is more like the "how") or would we only like to refer to what it makes possible? 

```suggestion
- **Extension point**: A place in Grafana Core or in a plugin where content can be hooked into.
```

> - **Renderable content**: Functionality (link or component) made available to render in an extension point. To make content available to an extension point it must be either exposed or registered.

Reviewer: ```suggestion
- **Renderable content**: Functionality (link or component) made available to render in an extension point or to be exposed.
```

> - **Exposed component**: Component made available with the `expose*` APIs. If exposed, content is free to use without restrictions.

Reviewer: ```suggestion
- **Exposed component**: Component made available to other plugins with the `expose*` APIs. If exposed, a component is not tied to extension points.
```

> - **Registered content**: Links or components made available with `the add*` APIs. If registered, content can be rendered in specific extension points..

Reviewer: ```suggestion
- **Registered content**: Links or components made available with `the add*` APIs. If registered, content can be rendered in specific extension points.
```

- **Content provider**: The app plugin providing the content (link or component) to be rendered at the extension point. 

- **Content user**: The extension point using the renderable content.

- **Plugin developer**: A developer working with the Grafana plugins ecosystem.

## I want to render extension content 

> As a content user, you need to define an extension point in your plugin to be able to render content (links or components) made available by content providers.

Reviewer: Should we mention "exposed components" here? (Which don't require an extension point.)

Author: Yes!

### Why add an extension point?

Define extension points to add new capabilities:

- An extension point allows other plugins to extend your UI with new functionality. You don't need any additional effort to provide functionality from other plugins in your UI.
- Clean separation of concerns. Your application doesn't need to know anything about the plugin extending your UI.
- Easy to bootstrap. If both apps are installed and enabled, then the extensions are automatically configured and displayed for your user. There is no need for either app to include custom logic to detect the presence of the other.

> ### Where can I find the content rendered in my extension point?

Reviewer: Maybe I am misunderstanding something, and in that case I am sorry. To me this heading ("Where can I find the content rendered in my extension point?") refers to how I could locate my extension point with any registered extensions on the UI, while below we are showing an example for an extension point. Is that intended?
